NodeInstaller fails at renameTo Node Binary with existing NodeJS version (Windows)

**Do you want to request a _feature_ or report a _bug_?**
bug

**What is the current behavior?**
The [NodeInstaller.java](https://github.com/eirslett/frontend-maven-plugin/blob/master/frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/NodeInstaller.java#L154) fails to update an existing NodeJS version (Node.exe) using the renameTo function. The task fails. If I delete the node folder manually before running the task, the NodeJS executable is installed properly.

**If the current behavior is a bug, please provide the steps to reproduce.**

[INFO] --- frontend-maven-plugin:1.4:install-node-and-npm (install node and npm) @ PROJECT ---
[INFO] Node v6.12.0 was installed, but we need version v8.9.1
[INFO] Installing node version v8.9.1
[INFO] Downloading https://CORPORATE-PROXY/nodejs/v8.9.1/node-v8.9.1-win-x64.zip to C:\m2_repo\com\github\eirslett\node\8.9.1\node-8.9.1-win-x64.zip
[INFO] No proxies configured
[INFO] No proxy was configured, downloading directly
[INFO] Unpacking C:\m2_repo\com\github\eirslett\node\8.9.1\node-8.9.1-win-x64.zip into C:\PATH\TO\PROJECT\node\tmp
[INFO] Copying node binary from C:\PATH\TO\PROJECT\node\tmp\node-v8.9.1-win-x64\node.exe to C:\PATH\TO\PROJECT\node\node.exe
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:25 min
[INFO] Finished at: 2017-11-17T10:01:40+01:00
[INFO] Final Memory: 66M/495M
[INFO] ------------------------------------------------------------------------
**[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.4:install-node-and-npm (install node and npm) on project PROJECT: Could not install Node: Was not allowed to rename C:\PATH\TO\PROJECT\node\tmp\node-v8.9.1-win-x64\node.exe to C:\PATH\TO\PROJECT\node\node.exe -> [Help 1]**

**What is the expected behavior?**
Either one of:
- renameTo works as expected
- or: Try to delete any existing executable file in advance, before copying over a new version instead of using renameTo

**Please mention your frontend-maven-plugin and operating system version.**
Plugin: 1.4 (haven't seen any relevant changes in the commits of 1.5 and 1.6)
OS: Windows 7 x64

Best Regards,
Jan

I encountered this issue as well. For us the build was running on a windows Jenkins CI server with Maven 3.x and Plugin 1.6, and I think Windows 10 Enterprise x64..

The issue appears to be because of the existing workspace containing the previous node version with the folder name as _..blah\..\node_. So the new version that is downloaded cannot be renamed to this existing folder.
**Workaround** - Specify (check) the _Delete workspace before Build_ option for 1 build cycle, so that the workspace is deleted and the plugin downloads the Node version and renames it without any issue since no existing node folder is present. After build success, don't forget to un-check this option for faster builds.

I am giving some steps to reproduce the same, in case this wasn't clear from earlier posts:
**How to reproduce the same**
 - Assuming you have a _pom.xml_ with the _install-node-and-npm_ and _nodeVersion_ tag set to say v8.11.1 like:
    ```
<build>
        <plugins>
            <plugin>
                <groupId>com.github.eirslett</groupId>
                <artifactId>frontend-maven-plugin</artifactId>
                <executions>
                    <execution>
                        <id>install node and npm</id>
                        <goals>
                            <goal>install-node-and-npm</goal>
                        </goals>
                        <configuration>
                            <nodeVersion>v8.11.1</nodeVersion>                          
                            <workingDirectory>${basedir}</workingDirectory>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
     </build>
```
 - Delete the node folder in the project folder if present. 
- Run a _mvn clean install_ this will download the node version to the folder called _node_ and do other stuff as specified in your pom.xml
- Change the _nodeVersion_ to say v8.9.1 in the pom.xml
- Do a _mvn clean install_ again, you should get this same error

Error:
[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.6:install-node-and-npm (install node and npm) on project inline-window-ui: Could not install Node: Was not allowed to rename C:\dev\blah\ui\node\tmp\node-v8.9.1-win-x64\node.exe to C:\dev\blah\ui\node\node.exe -> [Help 1]
[ERROR]

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,10 @@ Last public release: [![Maven Central](https://maven-badges.herokuapp.com/maven-
 
 ## Changelog
 
+### 1.7.6
+
+* Fix #670: Plugin will no longer fail to install node.exe if node.exe already exists 
+
 ### 1.5
 
 * Revert support for the maven.frontend.failOnError flag ([#572](https://github.com/eirslett/frontend-maven-plugin/pull/572)), due to

--- a/frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/NodeInstaller.java
+++ b/frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/NodeInstaller.java
@@ -3,6 +3,8 @@ package com.github.eirslett.maven.plugins.frontend.lib;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
 import java.util.Arrays;
 
 import org.apache.commons.io.FileUtils;
@@ -151,7 +153,12 @@ public class NodeInstaller {
 
                 File destination = new File(destinationDirectory, "node");
                 this.logger.info("Copying node binary from {} to {}", nodeBinary, destination);
-                if (!nodeBinary.renameTo(destination)) {
+                if (destination.exists() && !destination.delete()) {
+                    throw new InstallationException("Could not install Node: Was not allowed to delete " + destination);
+                }
+                try {
+                    Files.move(nodeBinary.toPath(), destination.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                } catch (IOException e) {
                     throw new InstallationException("Could not install Node: Was not allowed to rename "
                         + nodeBinary + " to " + destination);
                 }
@@ -219,7 +226,9 @@ public class NodeInstaller {
 
                 File destination = new File(destinationDirectory, "node.exe");
                 this.logger.info("Copying node binary from {} to {}", nodeBinary, destination);
-                if (!nodeBinary.renameTo(destination)) {
+                try {
+                    Files.move(nodeBinary.toPath(), destination.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                } catch (IOException e) {
                     throw new InstallationException("Could not install Node: Was not allowed to rename "
                         + nodeBinary + " to " + destination);
                 }

